File: HW6 - C++/gtuset.cpp
#include <iostream>
#include "gtuset.h"

namespace EmirUzun{
	template <class T>
	GTUSet<T>::GTUSet(){
		this->usedSize=0;
		this->capacity = 10;
		std::shared_ptr<T> temp(new T[this->capacity]);
		this->container = temp;
	}


	template <class T >
	bool GTUSet<T>::empty()const{
        if(this->usedSize == 0) 	return true;
        else return false;
	}

	template <class T>
	int GTUSet<T>::size()const{
		return this->usedSize;
	}

	template <class T>
	void GTUSet<T>::insert(const T& element){

		for(auto it = this->begin(); it != this->end(); ++it){
			if(*(it) == element){
				throw std::invalid_argument("This element is already in the set container ! ");
				return;
			}
		}

		if(this->usedSize != this->capacity){
			*(this->end()) = element;
			std::cout << "'" << element << "' was inserted to set container ! " << std::endl; 
			this->usedSize++;
		}
		else if(this->usedSize == this->capacity){
			this->capacity += 10;
			std::shared_ptr<T> temp(new T[this->capacity]);
			int index = 0;
			for(auto it = this->begin() ; it != this->end();++it){
				temp.get()[index] = *(it);
				index++;
			}
			this->container = nullptr;
			this->container = temp;
			*(this->end()) = element;
			std::cout << "'" << element << "' was inserted to set container ! " << std::endl; 
			this->usedSize++;		
		}
	}

	template <class T>
	void GTUSet<T>::clear(){
		this->usedSize = 0;
	}

	template <class T>
	void GTUSet<T>::print(){
		
	}

	template <class T>
	void GTUSet<T>::erase(const T& element){
		for(auto it = this->begin(); it != this->end();++it){
			if(*(it) == element){
				for(auto it2 = it;it2 != this->end();++it2){
					*(it2) = *(it2.getPtr()+1);
				}
				std::cout << "'" << element << "' was erased ! " << std::endl;
				this->usedSize--;
				return;
			}
		}
		std::cout << "This element '"  << element << "' was not found ! " << std::endl;
	}

	template <class T>
	GTUIterator<T> GTUSet<T>::begin()const{
		return (this->container).get();
	}

	template <class T>
	GTUIterator<T> GTUSet<T>::end()const{
		GTUIterator<T> it = (this->container).get() ;
		for(int i = 0;i<this->usedSize;++i){
			it++;
		}
		return it;
	}

	template <class T> 
	int GTUSet<T>::max_size(){
		return this->capacity;
	}

}

File: HW8 - JAVA/src/GTUSet.java
import java.util.NoSuchElementException;
import java.lang.reflect.Array;
import java.security.InvalidParameterException;

/**
 * The Set Class which is extends from GTUContainer class.
 * @author Emirhan Uzun 171044019
 *
 */
public class GTUSet<T> extends GTUContainer<T> {
	
	/**
	 * @param classtype  The type of the container
	 * @param givencapacity This is capacity that is given from user.
	 */
	public GTUSet(Class<T[]> classtype, int givencapacity) {
		super(classtype, givencapacity);
	}
	
	
	/**
	 * This is the inner class that in the GTUSet class.It extends the GTUIterator methods and assign index.
	 *
	 */
	public class GTUSetIter extends GTUIterator<T>{
		int index = 0;
		
		@Override
		public boolean hasNext(){
	        if(index < usedSize ) return true;
	        return false;
	    }

	    
	    public T next() 
	    {
	        try{
	            if(!hasNext())
	                throw new NoSuchElementException("This is an exception that array has no more element");
	            return container[index++];
	        }
	        catch(Exception e){
	            System.out.println(e);
	            return null;
	        }
	    }


		@Override
		public int getIndex() {
			return index;
		}
	}

	@Override
	public boolean empty() {
		if(this.usedSize == 0) 	return true;
		else return false;
	}

	@Override
	public int size() {
		return this.usedSize;
	}

	@Override
	public int max_size() {
		return this.capacity;
	}

	/**
	 *@throws Insert element, throws exception java.lang.IllegalArgumentException if there is a problem with
insertion
	 */
	@Override
	public void insert(T element) throws InvalidParameterException {
			boolean flag = false;
			GTUIterator<T> it = this.iterator();
			while(it.hasNext()) {
				T val = it.next();
				if(val.equals(element) )
					flag = true;
			}
			if(!flag) {
				if(usedSize + 1 == capacity) {
					capacity += 10;
					T[] tempArr = classtype.cast(Array.newInstance(classtype.getComponentType(), capacity ));
					for(int i=0; i<usedSize; i++)
						tempArr[i] = container[i];
					container = classtype.cast(Array.newInstance(classtype.getComponentType(), capacity ));
					for(int i=0; i<usedSize; i++)
						container[i] = tempArr[i];
				}
				container[usedSize] = element;
				usedSize++;
				System.out.println(element+" was inserted the set container ! ");
			} else if (flag) {
				throw new InvalidParameterException("In set container, this element is already in that.");
			}
	}

	@Override
	public void erase(T element) {
		boolean flag = false;
		GTUIterator<T> it = iterator();
		while(it.hasNext()) {
			T val = it.next();
			if(val.equals(element)) {
				flag = true;
				break;
			}
		}
		if(flag) {
			for(int i=it.getIndex(); i< usedSize; i++)
				container[i-1] = container[i];
			usedSize--;
			System.out.println(element+" was erased from the set container ! ");
		}
		else System.out.println(element+" is not in the set !");
		
	}

	@Override
	public void clear() {
		usedSize = 0;
		
	}

	@Override
	public GTUIterator<T> iterator() {
		return new GTUSetIter();
	}
	
	/**
	 *This method prints the container 
	 */
	@Override
    public String toString(){
        String printArr = "";
        printArr += "\n";
        GTUIterator<T> iter = this.iterator();

        while(iter.hasNext()) {
                T element = iter.next();
                printArr += element;
                printArr += " ";
        }
        printArr += "\n";
        return printArr;
    }

	@Override
	public boolean contains(Object element) {
		GTUIterator<T> it = this.iterator();
		while(it.hasNext()) {
			T value = it.next();
			if(value == element) {
				return true;
			}
		}
		return false;
	}	

}
